include/fsm/walk.h

```diff
@@ -75,5 +75,45 @@ fsm_walk_edges(const struct fsm *fsm, void *opaque,
 	int (*callback_literal)(const struct fsm *, fsm_state_t, fsm_state_t, char c, void *),
 	int (*callback_epsilon)(const struct fsm *, fsm_state_t, fsm_state_t, void *));
 
+/* Walk a DFA and generate inputs that reach an end state.
+ * An internal buffer will be resized on demand. This generates
+ * matches depth-first, bounded by max_length.
+ *
+ * This does not produce every possible matching input string, only
+ * strings representing distinct paths through the DFA's graph. For
+ * example, when given the regex /^x[aeiou]y$/ it may only generate
+ * "xay", not "xey", "xiy", "xoy", or "xuy", because those are other
+ * labels for the same graph edge. This function was originally added
+ * to compare matching behavior against PCRE, so eliminating those
+ * functionally equivalent cases makes testing dramatically faster,
+ * but exploring every edge could be added later.
+ *
+ * Note: fsm is non-const because it calls fsm_trim on the FSM
+ * internally. This records the shortest distance from each state to an
+ * end state, which is used to prune branches that would not produce
+ * any new matches with the remaining buffer space. This makes generation
+ * dramatically faster.
+ *
+ * Returns 1 if iteration completes (or is halted by the callback),
+ * or 0 on error.
+ */
+enum fsm_generate_matches_cb_res {
+	/* Continue generating. */
+	FSM_GENERATE_MATCHES_CB_RES_CONTINUE,
+	/* Continue generating, but treat the current end state as a
+	 * dead end. */
+	FSM_GENERATE_MATCHES_CB_RES_PRUNE,
+	/* Halt iteration entirely. Clean up and return. */
+	FSM_GENERATE_MATCHES_CB_RES_HALT
+};
+typedef enum fsm_generate_matches_cb_res
+fsm_generate_matches_cb(const struct fsm *fsm,
+    size_t depth, size_t match_count, size_t steps,
+    const char *input, size_t input_length,
+    fsm_state_t end_state, void *opaque);
+int
+fsm_generate_matches(struct fsm *fsm, size_t max_length,
+    fsm_generate_matches_cb *cb, void *opaque);
```

Comment:
In `re` as well? Should I do that in this PR or separately?

I think adding a random walk for generation will be good, the existing graphs can obviously get pretty cyclic, but that seems like it belongs in its own PR -- this one is just about clearing out some misc. stuff that would make the PR for captures even bigger.

Reply:
There's a `-W <arg>` interface for `gen_words` that is not yet implemented; I could use that for "walk", since `-g` and `-G` are already taken.

On the other hand, `-G` is for `glushkovise`, which is now an alias for `-E` / `remove_epsilons`, and `-G` is available in `re`, so if I could use that for generating output in both, but that's a breaking change. It doesn't look like anything in `tests/` is using `-G` though. Thoughts?